### PR TITLE
Allow Panache bytecode enhancers to benefit from class transformers caches

### PR DESCRIPTION
**Pull request by Sanne**
See also: https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Bytecode.20transformers.20caches.20for.20Hibernate.20entities

### CODE REVIEW

File: extensions/panache/panache-hibernate-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheHibernateCommonResourceProcessor.java [hunk 1]
```diff
@@ -126,7 +126,7 @@ void replaceFieldAccesses(CombinedIndexBuildItem index,
                 }
                 produced.add(cn);
                 transformers.produce(
-                        new BytecodeTransformerBuildItem(cn, panacheFieldAccessEnhancer, entityClassNamesInternal));
+                        new BytecodeTransformerBuildItem(true, cn, panacheFieldAccessEnhancer, entityClassNamesInternal, true));
```
Comment (yrodiere) [inline extensions/panache/panache-hibernate-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheHibernateCommonResourceProcessor.java:129]: FWIW the first `true` is making the transformer eager, whereas it previously was lazy.

Not sure this changes anything though... It seems this information is not used anywhere. It cascades down to this code, which writes to the set but never reads from it :x 

https://github.com/quarkusio/quarkus/blob/3374799f6d0523767d978ecbd35dbbe96f258ad5/core/deployment/src/main/java/io/quarkus/runner/bootstrap/StartupActionImpl.java#L63-L64

Comment (yrodiere) [reply extensions/panache/panache-hibernate-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheHibernateCommonResourceProcessor.java:129]: As I mentioned, we're not keeping them lazy. The boolean is ultimately used to populate a `Set` which nobody ever reads from.

Comment (Sanne) [reply extensions/panache/panache-hibernate-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheHibernateCommonResourceProcessor.java:129]: ho... fun :)

File: extensions/panache/panache-hibernate-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheHibernateCommonResourceProcessor.java [hunk 2]
```diff
@@ -125,8 +130,13 @@ void replaceFieldAccesses(CombinedIndexBuildItem index,
                     continue;
                 }
                 produced.add(cn);
-                transformers.produce(
-                        new BytecodeTransformerBuildItem(cn, panacheFieldAccessEnhancer, entityClassNamesInternal));
+                final BytecodeTransformerBuildItem transformation = new BytecodeTransformerBuildItem.Builder()
+                        .setClassToTransform(cn)
+                        .setCacheable(true)
```
[the remark this answers is not in the document]
Comment (Sanne) [reply extensions/panache/panache-hibernate-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheHibernateCommonResourceProcessor.java:135]: Right, that's precisely what I'm not sure of in this case - I believe it might be safe in this particular case due to the nature of the transformation but I'd really like @FroMage to have a look first. (I asked his review)

[the remark this answers is not in the document]
Comment (FroMage) [reply extensions/panache/panache-hibernate-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheHibernateCommonResourceProcessor.java:135]: Everything is cacheable with the proper cache eviction and invalidation. How is that determined here? I've never used transformer caches.

**Comment by Sanne** [comment]
"Undrafted" and rebased, let's have CI opinions as well.

I described the rationale above and think it should be safe after having convinced myself while writing the rationale :) but let's have @FroMage review it please.

File: extensions/panache/panache-hibernate-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheHibernateCommonResourceProcessor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (yrodiere) [reply extensions/panache/panache-hibernate-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheHibernateCommonResourceProcessor.java:135]: > * when instrumenting the client side, it's a bit more tricky as it literally calls code on the entity; but I believe that it would be safe as well as the client - when unmodified from a previous version of the app - can't possibly be writing on a field which was just introduced, or removed, as it wouldn't have compiled

I'm afraid this is a bit optimistic. We're not talking about class fields being added/removed here, we're talking about JPA attributes being added/removed. So I can think of several situations where a client would already be accessing a class field, and would require re-transformation because of a change that lies exclusively in the accessed class:

* the accessed class was missing a `@Entity` annotation, and that annotation was just added
* the accessed field was annotated with `@Transient`, and that annotation gets removed
* the accessed field was NOT annotated with `@Transient`, and that annotation gets added
* and probably many more

So IMO making this transformer cacheable, as things are, is unsafe.

This could of course be made safe by making sure that classes accessing an entity get their cache entry evicted when that entity is modified (or its "entity" status is added/removed). But I suspect that's a lot more work than it's worth?

[the remark this answers is not in the document]
Comment (FroMage) [reply extensions/panache/panache-hibernate-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheHibernateCommonResourceProcessor.java:135]: I agree with @yrodiere, here you want to cache use-site transformations. The use-site transformations depend not on the current use-site class according to its `ClassChangeInformation` but on the set of entities declared, and their set of JPA attributes. So the cache invalidation is wrong, unless I'm missing something, and so this change is dangerous, and it should be easy to test that by testing any of the scenarios that @yrodiere described.

To sum it up, when an entity changes (anywhere in its jpa-mapped hierarchy), the set of field-access transformers to invalidate is the set of classes who are using this entity (and its hierarchy), as reported by Jandex's use-site indexing.

**Review by FroMage — DISMISSED** [review]

File: extensions/panache/panache-hibernate-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheHibernateCommonResourceProcessor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Sanne) [reply extensions/panache/panache-hibernate-common/deployment/src/main/java/io/quarkus/panache/common/deployment/PanacheHibernateCommonResourceProcessor.java:135]: Thank you all, that's some great examples of when it would fail - and I hate the complexity deriving from a "bad cache", that would introduce some seriously obscure bugs.

I won't be refactoring the whole cache invalidation strategy to accommodate for this as I don't have the bandwith for now - but we might want to get back to that eventually if this is a problem.

Do we agree the other transformation above is safe though? The one applied on entities?

**Comment by Sanne** [comment]
Updated & rebased

**Review by FroMage — APPROVED** [review]
Yeah, that seems safer, I suppose transforming the accessors is only local-class dependent.

**Comment by Sanne** [comment]
Pushed a follow-up as #40629
